### ISSUE TITLE
Chat stuttering

### ISSUE THREAD

**Issue by jbob182** [NodeBB/NodeBB#3558]
We use the chat feature quite often when working with our user base. However, the chat feature can become unusable at times due to "stuttering". This can be observed in two ways:

1.) Type a message and press enter repeated quickly. You'll notice the message is sent multiple times.
2.) Type a message, hit enter and begin typing a new sentence right away. The second sentence will get cut off.

Why?
====================
I can only speculate as I have not looked at the code but here are my thoughts:

The chat system is a simple client <-> server <-> client setup, where a client tells the server the new message and the server responds "ok" or something similar. 

The problem is if the server is under a heavy load (when you have a large user base) where the socket response from the chat server is more than a few milliseconds. So the message stays in a 1 entity queue. When the "ok" response is received, it clears the entire text area for typing. Alternatively, if repeatedly pressing enter, it sends the same message over and over to the server, not waiting for the "ok" response.

Proposed Solution
====================
While not the only fix, one solution that address both scenarios provided is to use a queue. When the user presses enter, immediately put the message into the queue and clear the text area without waiting for the "ok". Multiple entities should be able to be placed in said queue (in case of sever server load). When the server responds "ok" to a message, pop the queue.

**Comment by jbob182** [comment]
After examining the Chats.sendMessage code (https://github.com/NodeBB/NodeBB/blob/master/public/src/client/chats.js#L186)  a simple patch for this issue would just be to move ```inputEl.val('');```  outside of the callback

**Comment by psychobunny** [comment]
Pretty much. We're waiting on an author who created a pull request who hasn't signed the CLA yet. But I did enjoy the 1000 line writeup on how to fix this... :P kidding 

Anyways, here's the link: https://github.com/NodeBB/NodeBB/pull/3532

I could be wrong but since you've discovered this yourself and if you submit a PR with the signed CLA we can merge yours and close the other. Or just wait for original pull author to respond

**Comment by julianlam** [comment]
Duplicate of #2427
